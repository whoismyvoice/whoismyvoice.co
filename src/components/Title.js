import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import '../styles/TitleComponent.css';

export class Title extends Component {
  static defaultProps = {
    className: '',
    templateData: {},
    templateString: '',
  };

  static propTypes = {
    className: PropTypes.string,
    templateData: PropTypes.shape({}),
    templateString: PropTypes.string,
  };

  constructor(props) {
    super(props);
    this.template = _.template(props.templateString);
  }

  componentWillUpdate(nextProps, nextState, nextContext) {
    if (
      this.state !== undefined &&
      this.state.templateString !== undefined &&
      this.state.templateString !== nextState.templateString
    ) {
      this.template = _.template(nextState.templateString);
    } else if (this.props.templateString !== nextProps.templateString) {
      this.template = _.template(nextProps.templateString);
    }
  }

  getTemplateData() {
    if (this.state && this.state.templateData !== undefined) {
      return this.state.templateData;
    } else if (this.props.templateData !== undefined) {
      return this.props.templateData;
    } else {
      return {};
    }
  }

  render() {
    const { className } = this.props;
    const content = { __html: this.template(this.getTemplateData()) };
    return (
      <div className={`title-component ${className}`}>
        <div
          className="title-component__description"
          dangerouslySetInnerHTML={content}
        />
      </div>
    );
  }
}

export default Title;
