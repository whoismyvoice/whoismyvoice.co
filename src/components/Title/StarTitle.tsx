import React, { FC } from 'react';

import '../../styles/TitleComponent.scss';

const Stars: FC<unknown> = () => (
  <div className="three-stars">
    <span>&#9733;</span>
    <span>&#9733;</span>
    <span>&#9733;</span>
  </div>
);

export const StarTitle: FC<unknown> = ({ children }) => (
  <div className="title-component uppercase">
    <Stars />
    <div className="title-component__description">{children}</div>
  </div>
);

export default StarTitle;
