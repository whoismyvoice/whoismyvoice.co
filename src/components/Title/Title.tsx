import * as React from 'react';
import * as _ from 'lodash';

import '../../styles/TitleComponent.css';

export interface Props {
  className?: string;
  templateData?: object;
  templateString?: string;
}

export interface State {
  templateData?: object;
  templateString?: string;
}

export class Title<P extends Props, S extends State> extends React.Component<
  P,
  S
> {
  static defaultProps = {
    className: '',
    templateData: {},
    templateString: '',
  };

  template: _.TemplateExecutor;

  constructor(props: P) {
    super(props);
    this.template = _.template(props.templateString);
  }

  componentWillUpdate(nextProps: Readonly<P>, nextState: Readonly<S>) {
    if (
      this.state !== undefined &&
      this.state.templateString !== undefined &&
      this.state.templateString !== nextState.templateString
    ) {
      this.template = _.template(nextState.templateString);
    } else if (this.props.templateString !== nextProps.templateString) {
      this.template = _.template(nextProps.templateString);
    }
  }

  getTemplateData() {
    if (this.state && this.state.templateData !== undefined) {
      return this.state.templateData;
    } else if (this.props.templateData !== undefined) {
      return this.props.templateData;
    } else {
      return {};
    }
  }

  render() {
    const { className } = this.props;
    const content = { __html: this.template(this.getTemplateData()) };
    return (
      <div className={`title-component ${className}`}>
        <div
          className="title-component__description"
          dangerouslySetInnerHTML={content}
        />
      </div>
    );
  }
}

export default Title;
