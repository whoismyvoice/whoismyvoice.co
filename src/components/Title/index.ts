import { Title as Component } from './Title';

export default Component;
