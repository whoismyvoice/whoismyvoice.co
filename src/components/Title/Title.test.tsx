import * as React from 'react';
import * as renderer from 'react-test-renderer';
import Title from './Title';

jest.mock('mixpanel-browser');

it('renders correctly no props', () => {
  const tree = renderer.create(<Title />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with className', () => {
  const tree = renderer.create(<Title className="foo" />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with templateData', () => {
  const data = {
    foo: 'bar',
  };
  const tree = renderer.create(<Title templateData={data} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with templateString', () => {
  const templateString = `My template string.`;
  const tree = renderer
    .create(<Title templateString={templateString} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with templateData & templateString', () => {
  const templateString = `My template string <%= foo %>.`;
  const data = {
    foo: 'bar',
  };
  const tree = renderer
    .create(<Title templateData={data} templateString={templateString} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
