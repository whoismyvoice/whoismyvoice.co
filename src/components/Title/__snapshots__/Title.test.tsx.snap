// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly no props 1`] = `
<div
  className="title-component "
>
  <div
    className="title-component__description"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;

exports[`renders correctly with className 1`] = `
<div
  className="title-component foo"
>
  <div
    className="title-component__description"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;

exports[`renders correctly with templateData & templateString 1`] = `
<div
  className="title-component "
>
  <div
    className="title-component__description"
    dangerouslySetInnerHTML={
      Object {
        "__html": "My template string bar.",
      }
    }
  />
</div>
`;

exports[`renders correctly with templateData 1`] = `
<div
  className="title-component "
>
  <div
    className="title-component__description"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;

exports[`renders correctly with templateString 1`] = `
<div
  className="title-component "
>
  <div
    className="title-component__description"
    dangerouslySetInnerHTML={
      Object {
        "__html": "My template string.",
      }
    }
  />
</div>
`;
